### ISSUE TITLE
Understanding training data

### ISSUE THREAD

**Issue by tonmoyborah** [affinelayer/pix2pix-tensorflow#138]
Hi, I know this is old and maybe the response isn't as prompt now but I'd like to ask you guys a few questions about how to improve my results . @christopherhesse  @julien2512 
I want to use this repo along with [face2face](https://github.com/datitran/face2face-demo) to learn from facial landmarks and translate this into a face. Basically, I want this network to learn how to recreate a person's face (say A) from facial landmarks. I first trained the network on frames from a single video (around 500 images for 200-500 epochs and other default settings) but the results were very bad. Then I added random images of A (very similar pose but different facial expressions) from google scrap and trained 700 images for 200-500 epochs. The results were a little better but still it didn't quite look like a person. I read about @julien2512 's explanations about losses and I feel it is somewhat clear to me and the GAN loss should start decreasing ideally. But in my case it has only been increasing (till 1000+ epochs on 700 images). What am I doing wrong? 
My question is, what kind of data does this network prefer? Is it better if I give images from a single video? Or multiple videos? Or should I provide diverse images and not only from videos (as frames are not very different in videos)?
How many images do you suggest I have? Is 700 very less for good results?
Should I  keep training for more epochs?
Also, another consideration is that if we provide images from very similar videos, it might be easier for the network to recreate a similar image? Since if we provide diverse images, the color of clothes, lighting,etc will change  and might make the network more confused. ..am I thinking correctly?
![tensorboard](https://user-images.githubusercontent.com/10391022/48344133-46ae9a00-e69a-11e8-994b-174e58ee0b37.png)

**Comment by davesean** [comment]
> But in my case it has only been increasing (till 1000+ epochs on 700 images)

The reason for this, if you haven't changed any parameters, is that the L1 weighting is set to 100. So at the moment 0.12 * 100 > 5. This leads to a decrease in L1 but increase in GAN error. If you want, you can try to set the L1 weighting to 10 or 5 and see what happens.
Also, someone mentioned it already in another post, but the momentum of the batch_normalization should be set to 0.9 instead 0.1. Not sure if this is effecting your network.
About all your other questions, I'm not sure at all as I am too struggling with this network.

**Comment by tonmoyborah** [comment]
Hi @davesean 
Thanks for the tip. By momentum, do you mean the momentum term in the adam optimizer that we can pass as argument in `pix2pixp.py` (it is set to default 0.5)? I cannot find another momentum term in this project.
Regards

**Comment by davesean** [comment]
No, no, I mean the momentum term that is hard coded in the batch normalisation.

`def batchnorm(inputs):`

`    return tf.layers.batch_normalization(inputs, axis=3, epsilon=1e-5, momentum=0.1, training=True, gamma_initializer=tf.random_normal_initializer(1.0, 0.02))`

This momentum term should be set to 0.9. Let me know if this changed anything in your output!

**Comment by tonmoyborah** [comment]
Actually I am using the tf 1.1 version (April 2017 commit) due to compatibility issues with [face2face repo](https://github.com/datitran/face2face-demo). I realized this and removed my comment. Looks like `momentum` was not hard coded into batchnorm layer in tf 1.1. 

    def batchnorm(input):
    with tf.variable_scope("batchnorm"):
        # this block looks like it has 3 inputs on the graph unless we do this
        input = tf.identity(input)

        channels = input.get_shape()[3]
        offset = tf.get_variable("offset", [channels], dtype=tf.float32, initializer=tf.zeros_initializer())
        scale = tf.get_variable("scale", [channels], dtype=tf.float32, initializer=tf.random_normal_initializer(1.0, 0.02))
        mean, variance = tf.nn.moments(input, axes=[0, 1, 2], keep_dims=False)
        variance_epsilon = 1e-5
        normalized = tf.nn.batch_normalization(input, mean, variance, offset, scale, variance_epsilon=variance_epsilon)
        return normalized

There is only one option for momentum and it is for the adam optimizer and can be passed as an argument (default 0.5). Maybe I will change it to 0.9 and check. Any pointers @julien2512 ?

**Comment by julien2512** [comment]
Most says something like 50 000 images to start a deep learning, and more if you can !

And you are right when you ask "Since if we provide diverse images, the color of clothes, lighting,etc will change and might make the network more confused. ..am I thinking correctly?".
You also need to choose a good representative set of the overall.
If you got metadata attach to each image, you can try to use the Embeddings from tensorboard to visualize it.

Knowing your data can lead in a better model to be learn !

**Comment by tonmoyborah** [comment]
Thanks for your reply @julien2512 .
So diverse images might be a problem, so I should stick to using frames from a single video. It is difficult/impossible to get better results with this. How do you suggest I get more training data? Also, does data augmentation work in GANs?

**Comment by julien2512** [comment]
May be this can help you.

If that person A do not give a smile, the neural network will never be able to guess it.

Instead of learning that smile, your network is probably learning things you cannot guess.
Like the distance between the mouth and the ear, or their respective proportions.
If you give it a try with a landmark with no respect of these distance and proportion, will the face still be the person A ? Will it still like a human ?

Reconsider your objective regarding your data :
`learn how to recreate a person's face (say A) from facial landmarks`
If the objective is to draw whatever smile and see the person A smile, your data is probably wrong : it would be better to use metadata instead.

Hoping that you find your happiness,

**Comment by tonmoyborah** [comment]
Thanks

**Comment by tonmoyborah** [comment]
@julien2512  whenever you have time, it'd be great if you could give your opinion on this [doubt](https://github.com/affinelayer/pix2pix-tensorflow/issues/32#issuecomment-441210260) that I had.
Thanks a lot

**Comment by mrgloom** [comment]
>The reason for this, if you haven't changed any parameters, is that the L1 weighting is set to 100. So at the moment 0.12 * 100 > 5. This leads to a decrease in L1 but increase in GAN error.

What is `0.12`?
